cmd/gopherbadgeimg/main.go

```diff
@@ -0,0 +1,301 @@
+	// Since we have a byte slice, and 8 bytes per bit, divide by 8
+	imageBits := make([]byte, x*y/8)
+
+	// Again, on or off, white or black are our only color options
+	palette := []color.Color{
+		color.Black,
+		color.White,
+	}
+
+	if disableDithering {
+		// don't dither image if flag is set, useful for some images which are already black and white
+	} else {
+		// using our palette, create a dithering struct
+		// and dither our image to get some false shading.
+		// read more here: https://en.wikipedia.org/wiki/Floyd%E2%80%93Steinberg_dithering
+		d := dither.NewDitherer(palette)
+		d.Matrix = dither.FloydSteinberg
+		dithered := d.Dither(dst)
+		// this nil check is necessary since the library will often write
+		// the dithered image to dst, but not always. Read their docs for more info
+		if dithered != nil {
+			var ok bool
+			dst, ok = dithered.(*image.RGBA)
+			// docs claim image is guaranteed to be of this type when not nil, but it's good to check anyway
+			if !ok {
+				log.Fatalf("error: typeof dithered should have been `*image.RGBA` but was `%T`", dithered)
+			}
+		}
+	}
+
+	// loop over the x axis first, then y as screen updates LTR, top to bottom
+	// (vertical axis must be inner loop) for the badge layout
+	for i := 0; i < x; i++ {
+		for j := 0; j < y; j++ {
+			// grab dithered image point, determine if bit should be 1 or a 0
+			r, g, b, _ := dst.At(x-1-i, j).RGBA()
+			if r+g+b == 0 {
+				// use bit shifting + integer division & modulo arithmetic to change
+				// the individual bits we want to set
+				//			imageBits[(i*y+j)/8] = imageBits[(i*y+j)/8] | (1 << uint(7-(i*y+j)%8))
+				index := (i*y + j) / 8
+				bit := uint(7 - (i*y+j)%8)
+				imageBits[index] |= (1 << bit)
+			}
+		}
+	}
+	return imageBits
+}
+
+func ParseRatio(rstr string) (int, int, error) {
+	rstr = strings.ToLower(rstr)
+	pixels := strings.Split(rstr, "x")
+	if len(pixels) != 2 {
+		return 0, 0, errors.New("invalid ratio string provided")
+	}
+	x, err := strconv.Atoi(pixels[0])
+	if err != nil {
+		return 0, 0, errors.Join(errors.New("error: could not parse x coordinate count"), err)
+	}
+	y, err := strconv.Atoi(pixels[0])
+	if err != nil {
+		return 0, 0, errors.Join(errors.New("error: could not parse y coordinate count"), err)
+	}
+	return x, y, nil
+}
+
+// Usage prints a proper example of usage for when the user misuses the program.
+//
+// Usage also calls exit(1) to terminate the program with an error code.
+func Usage() {
+	fmt.Fprintf(flag.CommandLine.Output(), "Usage of %s <input_image>:\n", os.Args[0])
+	flag.PrintDefaults()
+	fmt.Fprintf(
+		flag.CommandLine.Output(),
+		"\nExamples:\n%s input.png -outmode bin -ratio profile\n%s input.jpg -outmode rice -ratio 128x128 -disable-dithering -show\n",
+		os.Args[0],
+		os.Args[0],
+	)
+	os.Exit(1)
+}
+
+// PrintImg prints an `*` for each marked bit
+//
+// It writes to stderr so that it doesn't conflict with the base64 output
+func PrintImg(x, y int, imgBits []byte) {
```

Comment:
This is wrong if the image is not a square, the new line occurs at the wrong time

Reply:
Yes, I tested it with many dimensions and eyeballed it but my Atoi([0]) error made them all square

I've fixed it now, used a different method which is easier to understand, will attach images